src/driver/xen.rs

```diff
@@ -112,58 +114,133 @@ impl Introspectable for Xen {
+            cr2: hvm_cpu.cr2,
+            sysenter_cs: hvm_cpu.sysenter_cs,
+            sysenter_esp: hvm_cpu.sysenter_esp,
+            sysenter_eip: hvm_cpu.sysenter_eip,
+            msr_efer: hvm_cpu.msr_efer,
+            msr_star: hvm_cpu.msr_star,
+            msr_lstar: hvm_cpu.msr_lstar,
             cs: SegmentReg {
-                base: 0,
-                limit: 0,
-                selector: 0,
+                base: hvm_cpu.cs_base,
+                limit: hvm_cpu.cs_limit,
+                selector: hvm_cpu.cs_sel.try_into().unwrap(),
             },
             ds: SegmentReg {
-                base: 0,
-                limit: 0,
-                selector: 0,
+                base: hvm_cpu.ds_base,
+                limit: hvm_cpu.ds_limit,
+                selector: hvm_cpu.ds_sel.try_into().unwrap(),
             },
             es: SegmentReg {
-                base: 0,
-                limit: 0,
-                selector: 0,
+                base: hvm_cpu.es_base,
+                limit: hvm_cpu.es_limit,
+                selector: hvm_cpu.es_sel.try_into().unwrap(),
             },
             fs: SegmentReg {
-                base: 0,
-                limit: 0,
-                selector: 0,
+                base: hvm_cpu.fs_base,
+                limit: hvm_cpu.fs_limit,
+                selector: hvm_cpu.fs_sel.try_into().unwrap(),
             },
             gs: SegmentReg {
-                base: 0,
-                limit: 0,
-                selector: 0,
+                base: hvm_cpu.gs_base,
+                limit: hvm_cpu.gs_limit,
+                selector: hvm_cpu.gs_sel.try_into().unwrap(),
             },
             ss: SegmentReg {
-                base: 0,
-                limit: 0,
-                selector: 0,
+                base: hvm_cpu.ss_base,
+                limit: hvm_cpu.ss_limit,
+                selector: hvm_cpu.ss_sel.try_into().unwrap(),
             },
             tr: SegmentReg {
-                base: 0,
-                limit: 0,
-                selector: 0,
-            },
-            ldt: SegmentReg {
-                base: 0,
-                limit: 0,
-                selector: 0,
+                base: hvm_cpu.tr_base,
+                limit: hvm_cpu.tr_limit,
+                selector: hvm_cpu.tr_sel.try_into().unwrap(),
             },
         }))
     }
 
+    fn write_registers(&self, vcpu: u16, reg: Registers) -> Result<(), Box<dyn Error>> {
+        let (mut buffer, size) = self.xc.domain_hvm_getcontext(self.domid)?;
+        //println!("{}", size);
+        let mut offset: u32 = 0;
+        let hvm_save_cpu =
+            unsafe { mem::MaybeUninit::<__HVM_SAVE_TYPE_CPU>::zeroed().assume_init() };
+        let hvm_save_code_cpu: u16 = mem::size_of_val(&hvm_save_cpu.c).try_into().unwrap();
+        //let hvm_save_type_cpu = type_of(&hvm_save_cpu.t).try_into().unwrap();
+        let mut cpu: *mut hvm_hw_cpu = std::ptr::null_mut();
+        unsafe {
```

Comment:
You should never use unsafe blocks in libmicrovmi. The whole point of having crates like `xenctrl` or `kvmi` is to provide safe bindings. Anything that is unsafe should be done in there.